src/chemistry/mozart/mo_tuvx.F90

```diff
@@ -0,0 +1,190 @@
+                               mpicom
+
+!-----------------------------------------------------------------------
+! Local variables
+!-----------------------------------------------------------------------
+    class(core_t), pointer :: core
+    character, allocatable :: buffer(:)
+    type(string_t) :: config_path
+    integer :: pack_size, pos, i_core, i_err
+
+    config_path = tuvx_config_path
+    if( is_main_task ) call log_initialization( )
+
+#ifndef HAVE_MPI
+    call assert_msg( 113937299, is_main_task, "Multiple tasks present without " &
+                     //"MPI support enabled for TUV-x")
+#endif
+
+    ! construct a core on the primary process and pack it onto an MPI buffer
+    if( is_main_task ) then
+      core => core_t( config_path )
+      pack_size = core%pack_size( mpicom )
+      allocate( buffer( pack_size ) )
+      pos = 0
+      call core%mpi_pack( buffer, pos, mpicom )
+      deallocate( core )
+    end if
+
+    ! broadcast the core data to all MPI processes
+#ifdef HAVE_MPI
+    call mpi_bcast( pack_size, 1, MPI_INTEGER, main_task, mpicom, i_err )
+    if( .not. is_main_task ) allocate( buffer( pack_size ) )
+    call mpi_bcast( buffer, pack_size, MPI_CHARACTER, main_task, mpicom, i_err )
+#endif
+
+    ! unpack the core for each OMP thread on every MPI process
+    allocate( tuvx_ptrs( max_threads( ) ) )
+    do i_core = 1, size( tuvx_ptrs )
+    associate( tuvx => tuvx_ptrs( i_core ) )
+      allocate( tuvx%core_ )
+      pos = 0
+      call tuvx%core_%mpi_unpack( buffer, pos, mpicom )
+    end associate
+    end do
+
+  end subroutine tuvx_init
+
+!================================================================================================
+
+  subroutine tuvx_finalize( )
+!-----------------------------------------------------------------------
+!
+! Purpose: clean up memory associated with TUV-x calculators
+!
+!-----------------------------------------------------------------------
+
+!-----------------------------------------------------------------------
+! Local variables
+!-----------------------------------------------------------------------
+    integer :: i_core
+
+    if( allocated( tuvx_ptrs ) ) then
+      do i_core = 1, size( tuvx_ptrs )
+        if( associated( tuvx_ptrs( i_core )%core_ ) ) then
+          deallocate( tuvx_ptrs( i_core )%core_ )
+        end if
+      end do
+    end if
+
+  end subroutine tuvx_finalize
+
+!================================================================================================
+
+  integer function thread_id( )
+!-----------------------------------------------------------------------
+!
+! Purpose: returns the id of the current OpenMP thread, or 1 if not
+!          using OpenMP
+!
+!-----------------------------------------------------------------------
+#ifdef _OPENMP
+    use omp_lib,        only : omp_get_thread_num
+#endif
+
+#ifdef _OPENMP
+    thread_id = omp_get_thread_num( )
+#else
+    thread_id = 1
+#endif
```

Comment:
@fvitt - is this the right way to know if OpenMP is being used? It doesn't seem to be used in the modeling2 configuration I've been using, but I saw code similar to this in other parts of CAM